firedrake/slate/slate.py

```diff
@@ -296,6 +305,107 @@ def _key(self):
         return (type(self), self._function)
 
 
+class IndexedTensor(TensorBase):
+    """This class represents a tensor corresponding
+    to particular block of a mixed tensor.
+
+    :arg tensor: A block (mixed) tensor.
+    :arg idx: A tuple of indices denoting which block
+              to extract.
+    """
+
+    def __init__(self, tensor, idx):
+        """Constructor for the IndexedTensor class."""
+        if not isinstance(tensor, TensorBase):
+            raise TypeError("Can only index Slate expressions.")
+
+        if tensor.rank == 0:
+            raise ValueError("Cannot index scalars.")
+
+        assert len(idx) == tensor.rank, (
+            "Number of indices must match the tensor rank."
+        )
+
+        if not tensor.is_mixed:
+            raise ValueError("Can only index block (mixed) tensors")
```

Comment:
I think it makes sense to be able to index non-mixed (assuming that the indices are all 0).  This makes writing code outside much more natural (you don't have to special-case outside).  

You code do:

```python
def __new__(cls, tensor, idx):
    ...
    if not tensor.is_mixed:
        assert all(i == 0 for i in idx)
        return tensor
    return super().__new__(cls)
```